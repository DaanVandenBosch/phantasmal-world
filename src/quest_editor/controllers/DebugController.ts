import { Controller } from "../../core/controllers/Controller";
import { Property } from "../../core/observable/property/Property";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { GuiStore, GuiTool } from "../../core/stores/GuiStore";
import { LogEntry } from "../../core/Logger";
import { LogStore } from "../stores/LogStore";
import { Severity } from "../../core/Severity";
import { property } from "../../core/observable";

export class DebugController extends Controller {
    readonly can_debug: Property<boolean>;
    readonly can_step: Property<boolean>;
    readonly can_stop: Property<boolean>;
    readonly thread_ids: ListProperty<number>;
    readonly debugging_thread_id: Property<number | undefined>;
    readonly active_thread_id: Property<number | undefined>;
    readonly can_select_thread: Property<boolean>;
    readonly log: ListProperty<LogEntry>;
    readonly severity: Property<Severity>;

    constructor(
        gui_store: GuiStore,
        private readonly quest_editor_store: QuestEditorStore,
        private readonly log_store: LogStore,
    ) {
        super();

        this.can_debug = quest_editor_store.current_quest.map(q => q != undefined);

        this.can_step = quest_editor_store.quest_runner.paused;

        this.can_stop = quest_editor_store.quest_runner.running;

        this.thread_ids = quest_editor_store.quest_runner.thread_ids;

        this.debugging_thread_id = quest_editor_store.quest_runner.debugging_thread_id;

        this.active_thread_id = quest_editor_store.quest_runner.active_thread_id;

        this.can_select_thread = quest_editor_store.quest_runner.thread_ids.map(
            ids => ids.length > 0 && quest_editor_store.quest_runner.running.val,
        );

        this.log = log_store.log;
        this.severity = log_store.severity;

        this.disposables(
            gui_store.on_global_keydown(GuiTool.QuestEditor, "F5", this.debug),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F5", this.stop),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F6", this.resume),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F8", this.step_over),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F7", this.step_in),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F8", this.step_out),
        );
    }

    debug = (): void => {
        const quest = this.quest_editor_store.current_quest.val;

        if (quest) {
            this.quest_editor_store.quest_runner.run(quest);
        }
    };

    resume = (): void => {
        this.quest_editor_store.quest_runner.resume();
    };

    step_over = (): void => {
        this.quest_editor_store.quest_runner.step_over();
    };

    step_in = (): void => {
        this.quest_editor_store.quest_runner.step_into();
    };

    step_out = (): void => {
        this.quest_editor_store.quest_runner.step_out();
    };

    stop = (): void => {
        this.quest_editor_store.quest_runner.stop();
    };

    set_severity = (severity: Severity): void => {
        this.log_store.set_severity(severity);
    };

    select_thread = (thread_id: number): void => {
        this.quest_editor_store.quest_runner.set_debugging_thread(thread_id);
    };
}
