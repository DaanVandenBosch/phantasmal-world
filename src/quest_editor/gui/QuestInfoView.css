.quest_editor_QuestInfoView {
    box-sizing: border-box;
    padding: 3px;
    overflow: auto;
    outline: none;
}

.quest_editor_QuestInfoView table {
    width: 100%;
}

.quest_editor_QuestInfoView th {
    text-align: left;
}

.quest_editor_QuestInfoView .core_TextInput {
    width: 100%;
}

.quest_editor_QuestInfoView .core_TextArea {
    width: 100%;
}

.quest_editor_QuestInfoView textarea {
    width: 100%;
}
