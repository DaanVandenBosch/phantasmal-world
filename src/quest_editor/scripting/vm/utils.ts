export type BinaryNumericOperation = (a: number, b: number) => number;

export const numeric_ops: Record<
    "add" | "sub" | "mul" | "div" | "idiv" | "mod" | "and" | "or" | "xor" | "shl" | "shr",
    BinaryNumericOperation
> = {
    add: (a, b) => a + b,
    sub: (a, b) => a - b,
    mul: (a, b) => a * b,
    div: (a, b) => a / b,
    idiv: (a, b) => Math.floor(a / b),
    mod: (a, b) => a % b,
    and: (a, b) => a & b,
    or: (a, b) => a | b,
    xor: (a, b) => a ^ b,
    shl: (a, b) => a << b,
    shr: (a, b) => a >>> b,
};

export type ComparisonOperation = (a: number, b: number) => boolean;

/**
 * Short-circuiting fold.
 */
export function andfold<T, A>(
    fn: (acc: A, cur: T) => A | undefined,
    init: A,
    lst: T[],
): A | undefined {
    let acc = init;

    for (const item of lst) {
        const new_val = fn(acc, item);

        if (new_val === undefined) {
            return undefined;
        } else {
            acc = new_val;
        }
    }

    return acc;
}

/**
 * Short-circuiting reduce.
 */
export function andreduce<T>(fn: (acc: T, cur: T) => T | undefined, lst: T[]): T | undefined {
    return andfold(fn, lst[0], lst.slice(1));
}

/**
 * Applies the given arguments to the given function.
 * Returns the second argument if the function returns a truthy value, else undefined.
 */
export function andsecond<T>(fn: (first: T, second: T) => any, first: T, second: T): T | undefined {
    if (fn(first, second)) {
        return second;
    }
    return undefined;
}

export function rest<T>(lst: T[]): T[] {
    return lst.slice(1);
}
