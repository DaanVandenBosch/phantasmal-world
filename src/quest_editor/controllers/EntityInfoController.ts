import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { flat_map_to_list, list_property, property } from "../../core/observable";
import { Euler, Vector3 } from "three";
import { deg_to_rad } from "../../core/math";
import { TranslateEntityAction } from "../actions/TranslateEntityAction";
import { RotateEntityAction } from "../actions/RotateEntityAction";
import { euler } from "../model/euler";
import { entity_data } from "../../core/data_formats/parsing/quest/Quest";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { QuestEntityPropModel } from "../model/QuestEntityPropModel";

const DUMMY_VECTOR = Object.freeze(new Vector3());
const DUMMY_EULER = Object.freeze(new Euler());

export class EntityInfoController extends Controller {
    readonly unavailable: Property<boolean>;
    readonly enabled: Property<boolean>;
    readonly type: Property<string>;
    readonly name: Property<string>;
    readonly section_id: Property<string>;
    readonly wave: Property<string>;
    readonly wave_hidden: Property<boolean>;
    readonly position: Property<Vector3>;
    readonly rotation: Property<Euler>;
    readonly props: ListProperty<QuestEntityPropModel>;

    constructor(private readonly store: QuestEditorStore) {
        super();

        const entity = store.selected_entity;
        this.unavailable = entity.map(e => e == undefined);
        this.enabled = store.quest_runner.running.map(r => !r);

        this.type = entity.map(e => (e instanceof QuestNpcModel ? "NPC" : "Object"));
        this.name = entity.map(e => (e == undefined ? "" : entity_data(e.type).name));
        this.section_id = entity.flat_map(e =>
            e == undefined ? property("") : e.section_id.map(id => id.toString()),
        );
        this.wave = entity.flat_map(e =>
            e instanceof QuestNpcModel
                ? e.wave.flat_map(w => w?.id?.map(w => w.toString()) ?? property("None"))
                : property(""),
        );
        this.wave_hidden = entity.map(e => !(e instanceof QuestNpcModel));
        this.position = entity.flat_map(e => e?.position ?? property(DUMMY_VECTOR));
        this.rotation = entity.flat_map(e => e?.rotation ?? property(DUMMY_EULER));
        this.props = flat_map_to_list(e => e?.props ?? list_property(), entity);
    }

    focused = (): void => {
        this.store.undo.make_current();
    };

    set_pos_x(x: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const pos = entity.position.val;
            this.store.undo
                .push(
                    new TranslateEntityAction(
                        this.store,
                        entity,
                        entity.section.val,
                        entity.section.val,
                        pos,
                        new Vector3(x, pos.y, pos.z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_pos_y(y: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const pos = entity.position.val;
            this.store.undo
                .push(
                    new TranslateEntityAction(
                        this.store,
                        entity,
                        entity.section.val,
                        entity.section.val,
                        pos,
                        new Vector3(pos.x, y, pos.z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_pos_z(z: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const pos = entity.position.val;
            this.store.undo
                .push(
                    new TranslateEntityAction(
                        this.store,
                        entity,
                        entity.section.val,
                        entity.section.val,
                        pos,
                        new Vector3(pos.x, pos.y, z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_rot_x(x: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const rot = entity.rotation.val;
            this.store.undo
                .push(
                    new RotateEntityAction(
                        this.store,
                        entity,
                        rot,
                        euler(deg_to_rad(x), rot.y, rot.z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_rot_y(y: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const rot = entity.rotation.val;
            this.store.undo
                .push(
                    new RotateEntityAction(
                        this.store,
                        entity,
                        rot,
                        euler(rot.x, deg_to_rad(y), rot.z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_rot_z(z: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const rot = entity.rotation.val;
            this.store.undo
                .push(
                    new RotateEntityAction(
                        this.store,
                        entity,
                        rot,
                        euler(rot.x, rot.y, deg_to_rad(z)),
                        false,
                    ),
                )
                .redo();
        }
    }
}
