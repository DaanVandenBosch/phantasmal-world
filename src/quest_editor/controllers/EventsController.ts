import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestEventDagModel } from "../model/QuestEventDagModel";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { flat_map_to_list, list_property, map } from "../../core/observable";
import { QuestEventModel } from "../model/QuestEventModel";
import { EditEventDelayAction } from "../actions/EditEventDelayAction";
import { WaveModel } from "../model/WaveModel";
import { RemoveEventAction } from "../actions/RemoveEventAction";
import { CreateEventAction } from "../actions/CreateEventAction";

export class EventsController extends Controller {
    readonly event_dag: Property<QuestEventDagModel | undefined>;
    readonly event_sub_graphs: ListProperty<ListProperty<QuestEventModel>>;
    readonly enabled: Property<boolean>;
    readonly unavailable: Property<boolean>;

    constructor(private readonly store: QuestEditorStore) {
        super();

        this.enabled = store.quest_runner.running.map(r => !r);
        this.unavailable = store.current_quest.map(q => q == undefined);

        this.event_dag = map(
            (quest, area) => {
                if (quest && area) {
                    return quest.event_dags.get(area.id);
                } else {
                    return undefined;
                }
            },
            store.current_quest,
            store.current_area,
        );

        this.event_sub_graphs = flat_map_to_list(
            dag => dag?.connected_sub_graphs ?? list_property(),
            this.event_dag,
        );
    }

    focused = (): void => {
        this.store.undo.make_current();
    };

    is_selected(event: QuestEventModel): Property<boolean> {
        return this.store.selected_wave.map(selected => event.wave === selected);
    }

    add_event = (): void => {
        const quest = this.store.current_quest.val;
        const area = this.store.current_area.val;

        if (quest && area) {
            const event_dag = quest.get_event_dag_or_create(area.id);

            const parent = event_dag.events.find(
                event => event.wave === this.store.selected_wave.val,
            );

            const section_id =
                parent?.section_id ?? this.store.selected_entity.val?.section_id?.val ?? 1;

            const event_ids: number[] = [];
            const wave_ids: number[] = [];

            for (const event of event_dag.events) {
                event_ids.push(event.id);

                if (event.wave.section_id.val === section_id) {
                    wave_ids.push(event.wave.id.val);
                }
            }

            event_ids.sort((a, b) => a - b);
            wave_ids.sort((a, b) => a - b);

            // Find the first available wave id.
            let wave_id: number = 0;

            for (const existing_wave_id of wave_ids) {
                if (++wave_id !== existing_wave_id) {
                    break;
                }
            }

            if (wave_id === wave_ids.length) {
                wave_id++;
            }

            // Generate id.
            let id_str: string;

            if (parent) {
                // Generate id based on first ancestor id and amount of its child events.
                let ancestor = parent;

                while (true) {
                    // Always choose the first parent.
                    const [p] = event_dag.get_parents(ancestor);

                    if (p) {
                        ancestor = p;
                    } else {
                        break;
                    }
                }

                const sub_graph_size = event_dag.get_sub_graph(ancestor).length.val;
                id_str = `${ancestor.id}${sub_graph_size}`;
            } else {
                // Generate id based on section id and wave id.
                id_str = `${section_id}${wave_id}`;
            }

            let id = parseInt(id_str, 10);

            // Make sure id is unique.
            let existing_index: number = 0;

            while (true) {
                existing_index = event_ids.indexOf(id, existing_index);

                if (existing_index === -1) {
                    break;
                } else {
                    id++;
                }
            }

            this.store.undo
                .push(
                    new CreateEventAction(
                        quest,
                        event_dag,
                        new QuestEventModel(
                            id,
                            section_id,
                            new WaveModel(wave_id, area.id, section_id),
                            30,
                            0, // TODO: what is a sensible value for event.unknown?
                        ),
                        parent,
                    ),
                )
                .redo();
        }
    };

    remove_event = (event: QuestEventModel): void => {
        const quest = this.store.current_quest.val;
        const dag = this.event_dag.val;

        if (quest && dag) {
            this.store.undo.push(new RemoveEventAction(this.store, quest, dag, event)).redo();
        }
    };

    set_selected_wave = (wave?: WaveModel): void => {
        if (this.enabled.val) {
            this.store.set_selected_wave(wave);
        }
    };

    set_delay = (event: QuestEventModel, delay: number): void => {
        this.store.undo.push(new EditEventDelayAction(event, event.delay.val, delay)).redo();
    };
}
