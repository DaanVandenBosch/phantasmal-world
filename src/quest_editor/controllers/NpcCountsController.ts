import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { npc_data, NpcType } from "../../core/data_formats/parsing/quest/npc_types";
import { property } from "../../core/observable";

export type NameWithCount = { readonly name: string; readonly count: number };

export class NpcCountsController extends Controller {
    readonly npc_counts: Property<readonly NameWithCount[]>;
    readonly unavailable: Property<boolean>;

    constructor(store: QuestEditorStore) {
        super();

        this.unavailable = store.current_quest.map(q => q == undefined);

        this.npc_counts = store.current_quest
            .flat_map(quest => (quest ? quest.npcs : property([])))
            .map(this.update_view);
    }

    private update_view(npcs: readonly QuestNpcModel[]): NameWithCount[] {
        const npc_counts = new Map<NpcType, number>();

        for (const npc of npcs) {
            const val = npc_counts.get(npc.type) || 0;
            npc_counts.set(npc.type, val + 1);
        }

        const extra_canadines = (npc_counts.get(NpcType.Canane) || 0) * 8;

        // Sort by canonical order.
        const sorted_npc_counts = [...npc_counts].sort((a, b) => a[0] - b[0]);

        return sorted_npc_counts.map(([npc_type, count]) => {
            const extra = npc_type === NpcType.Canadine ? extra_canadines : 0;

            return { name: npc_data(npc_type).name, count: count + extra };
        });
    }
}
