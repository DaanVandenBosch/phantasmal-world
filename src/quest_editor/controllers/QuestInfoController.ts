import { Property } from "../../core/observable/property/Property";
import { EditIdAction } from "../actions/EditIdAction";
import { EditNameAction } from "../actions/EditNameAction";
import { EditShortDescriptionAction } from "../actions/EditShortDescriptionAction";
import { EditLongDescriptionAction } from "../actions/EditLongDescriptionAction";
import { QuestModel } from "../model/QuestModel";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Controller } from "../../core/controllers/Controller";

export class QuestInfoController extends Controller {
    readonly current_quest: Property<QuestModel | undefined>;
    readonly enabled: Property<boolean>;
    readonly unavailable: Property<boolean>;

    constructor(private readonly store: QuestEditorStore) {
        super();

        this.current_quest = store.current_quest;
        this.enabled = store.quest_runner.running.map(r => !r);
        this.unavailable = this.current_quest.map(q => q == undefined);
    }

    focused = (): void => {
        this.store.undo.make_current();
    };

    set_id = (id: number): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.store.undo.push(new EditIdAction(quest, quest.id.val, id)).redo();
        }
    };

    set_name = (name: string): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.store.undo.push(new EditNameAction(quest, quest.name.val, name)).redo();
        }
    };

    set_short_description = (short_description: string): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.store.undo
                .push(
                    new EditShortDescriptionAction(
                        quest,
                        quest.short_description.val,
                        short_description,
                    ),
                )
                .redo();
        }
    };

    set_long_description = (long_description: string): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.store.undo
                .push(
                    new EditLongDescriptionAction(
                        quest,
                        quest.long_description.val,
                        long_description,
                    ),
                )
                .redo();
        }
    };
}
