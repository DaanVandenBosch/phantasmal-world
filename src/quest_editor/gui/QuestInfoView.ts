import { ResizableWidget } from "../../core/gui/ResizableWidget";
import { el } from "../../core/gui/dom";
import { Episode } from "../../core/data_formats/parsing/quest/Episode";
import { NumberInput } from "../../core/gui/NumberInput";
import { Disposer } from "../../core/observable/Disposer";
import { TextInput } from "../../core/gui/TextInput";
import { TextArea } from "../../core/gui/TextArea";
import "./QuestInfoView.css";
import { UnavailableView } from "./UnavailableView";
import { QuestInfoController } from "../controllers/QuestInfoController";

export class QuestInfoView extends ResizableWidget {
    readonly element = el.div({ class: "quest_editor_QuestInfoView", tab_index: -1 });

    private readonly table_element = el.table();
    private readonly episode_element: HTMLElement;
    private readonly id_input = this.disposable(new NumberInput(0));
    private readonly name_input = this.disposable(
        new TextInput("", {
            max_length: 32,
        }),
    );
    private readonly short_description_input = this.disposable(
        new TextArea("", {
            max_length: 128,
            font_family: '"Courier New", monospace',
            cols: 25,
            rows: 5,
        }),
    );
    private readonly long_description_input = this.disposable(
        new TextArea("", {
            max_length: 288,
            font_family: '"Courier New", monospace',
            cols: 25,
            rows: 10,
        }),
    );

    private readonly unavailable_view = new UnavailableView("No quest loaded.");

    private readonly quest_disposer = this.disposable(new Disposer());

    constructor(ctrl: QuestInfoController) {
        super();

        const quest = ctrl.current_quest;

        this.table_element.append(
            el.tr({}, el.th({ text: "Episode:" }), (this.episode_element = el.td())),
            el.tr({}, el.th({ text: "ID:" }), el.td({}, this.id_input.element)),
            el.tr({}, el.th({ text: "Name:" }), el.td({}, this.name_input.element)),
            el.tr({}, el.th({ text: "Short description:", col_span: 2 })),
            el.tr({}, el.td({ col_span: 2 }, this.short_description_input.element)),
            el.tr({}, el.th({ text: "Long description:", col_span: 2 })),
            el.tr({}, el.td({ col_span: 2 }, this.long_description_input.element)),
        );

        this.bind_hidden(this.table_element, ctrl.unavailable);

        this.element.append(this.table_element, this.unavailable_view.element);

        this.element.addEventListener("focus", ctrl.focused, true);

        this.disposables(
            this.unavailable_view.visible.bind_to(ctrl.unavailable),

            quest.observe(({ value: q }) => {
                this.quest_disposer.dispose_all();

                this.episode_element.textContent = q ? Episode[q.episode] : "";

                if (q) {
                    this.quest_disposer.add_all(
                        this.id_input.value.bind_to(q.id),
                        this.id_input.value.observe(({ value }) => ctrl.set_id(value)),

                        this.name_input.value.bind_to(q.name),
                        this.name_input.value.observe(({ value }) => ctrl.set_name(value)),

                        this.short_description_input.value.bind_to(q.short_description),
                        this.short_description_input.value.observe(({ value }) =>
                            ctrl.set_short_description(value),
                        ),

                        this.long_description_input.value.bind_to(q.long_description),
                        this.long_description_input.value.observe(({ value }) =>
                            ctrl.set_long_description(value),
                        ),

                        this.enabled.bind_to(ctrl.enabled),
                    );
                }
            }),
        );

        this.finalize_construction();
    }

    protected set_enabled(enabled: boolean): void {
        super.set_enabled(enabled);

        this.id_input.enabled.val = enabled;
        this.name_input.enabled.val = enabled;
        this.short_description_input.enabled.val = enabled;
        this.long_description_input.enabled.val = enabled;
    }
}
