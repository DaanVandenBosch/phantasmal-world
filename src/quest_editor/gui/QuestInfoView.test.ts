import { QuestInfoController } from "../controllers/QuestInfoController";
import { undo_manager } from "../../core/undo/UndoManager";
import { QuestInfoView } from "./QuestInfoView";
import {
    create_area_store,
    create_quest_editor_store,
} from "../../../test/src/quest_editor/stores/store_creation";
import { with_disposer } from "../../../test/src/core/observables/disposable_helpers";
import { load_default_quest_model } from "../../../test/src/utils";

test("Renders correctly without a current quest.", () =>
    with_disposer(disposer => {
        const view = disposer.add(
            new QuestInfoView(
                disposer.add(new QuestInfoController(create_quest_editor_store(disposer))),
            ),
        );

        expect(view.element).toMatchSnapshot('should render a "No quest loaded." view');
    }));

test("Renders correctly with a current quest.", () =>
    with_disposer(async disposer => {
        const area_store = create_area_store(disposer);
        const store = create_quest_editor_store(disposer);
        const view = disposer.add(new QuestInfoView(disposer.add(new QuestInfoController(store))));

        await store.set_current_quest(load_default_quest_model(area_store));

        expect(view.element).toMatchSnapshot("should render property inputs");
    }));

test("When the view's element is focused the quest editor store's undo stack should become the current stack.", () =>
    with_disposer(async disposer => {
        const store = create_quest_editor_store(disposer);
        const view = disposer.add(new QuestInfoView(disposer.add(new QuestInfoController(store))));

        undo_manager.make_noop_current();

        view.element.focus();

        expect(undo_manager.current.val).toBe(store.undo);
    }));
