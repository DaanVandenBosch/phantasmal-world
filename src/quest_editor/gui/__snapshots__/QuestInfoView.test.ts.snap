// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with a current quest.: should render property inputs 1`] = `
<div
  class="quest_editor_QuestInfoView"
  tabindex="-1"
>
  <table>
    <tr>
      <th>
        Episode:
      </th>
      <td>
        I
      </td>
    </tr>
    <tr>
      <th>
        ID:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
          style="width: 54px;"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            min="0"
            step="1"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th>
        Name:
      </th>
      <td>
        <span
          class="core_TextInput core_Input"
        >
          <input
            class="core_TextInput_inner core_Input_inner"
            maxlength="32"
            type="text"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Short description:
      </th>
    </tr>
    <tr>
      <td
        colspan="2"
      >
        <div
          class="core_TextArea"
        >
          <textarea
            class="core_TextArea_inner"
            cols="25"
            maxlength="128"
            rows="5"
            style="font-family: \\"Courier New\\", monospace;"
          />
        </div>
      </td>
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Long description:
      </th>
    </tr>
    <tr>
      <td
        colspan="2"
      >
        <div
          class="core_TextArea"
        >
          <textarea
            class="core_TextArea_inner"
            cols="25"
            maxlength="288"
            rows="10"
            style="font-family: \\"Courier New\\", monospace;"
          />
        </div>
      </td>
    </tr>
  </table>
  <div
    class="quest_editor_UnavailableView"
    hidden=""
  >
    <label
      class="core_Label disabled"
    >
      No quest loaded.
    </label>
  </div>
</div>
`;

exports[`Renders correctly without a current quest.: should render a "No quest loaded." view 1`] = `
<div
  class="quest_editor_QuestInfoView"
  tabindex="-1"
>
  <table
    hidden=""
  >
    <tr>
      <th>
        Episode:
      </th>
      <td />
    </tr>
    <tr>
      <th>
        ID:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
          style="width: 54px;"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            min="0"
            step="1"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th>
        Name:
      </th>
      <td>
        <span
          class="core_TextInput core_Input"
        >
          <input
            class="core_TextInput_inner core_Input_inner"
            maxlength="32"
            type="text"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Short description:
      </th>
    </tr>
    <tr>
      <td
        colspan="2"
      >
        <div
          class="core_TextArea"
        >
          <textarea
            class="core_TextArea_inner"
            cols="25"
            maxlength="128"
            rows="5"
            style="font-family: \\"Courier New\\", monospace;"
          />
        </div>
      </td>
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Long description:
      </th>
    </tr>
    <tr>
      <td
        colspan="2"
      >
        <div
          class="core_TextArea"
        >
          <textarea
            class="core_TextArea_inner"
            cols="25"
            maxlength="288"
            rows="10"
            style="font-family: \\"Courier New\\", monospace;"
          />
        </div>
      </td>
    </tr>
  </table>
  <div
    class="quest_editor_UnavailableView"
  >
    <label
      class="core_Label disabled"
    >
      No quest loaded.
    </label>
  </div>
</div>
`;
